presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramJacknifeStateStrategy.java

```diff
@@ -0,0 +1,185 @@
+    {
+        histogram = other.histogram.clone();
+    }
+
+    public FixedHistogramJacknifeStateStrategy(SliceInput input)
+    {
+        histogram = new FixedDoubleBreakdownHistogram(input);
+    }
+
+    @Override
+    public void validateParameters(
+            long bucketCount,
+            double sample,
+            double weight,
+            double min,
+            double max)
+    {
+        throwIfIllegalParameters(
+                histogram.getBucketCount(),
+                histogram.getMin(),
+                histogram.getMax(),
+                bucketCount,
+                sample,
+                weight,
+                min,
+                max);
+    }
+
+    @Override
+    public void validateParameters(
+            long bucketCount,
+            double sample,
+            double weight)
+    {
+        throwIfIllegalParameters(
+                histogram.getBucketCount(),
+                histogram.getMin(),
+                histogram.getMax(),
+                bucketCount,
+                sample,
+                weight);
+    }
+
+    @Override
+    public void mergeWith(StateStrategy other)
+    {
+        histogram.mergeWith(((FixedHistogramJacknifeStateStrategy) other).histogram);
+    }
+
+    @Override
+    public void add(double value, double weight)
+    {
+        histogram.add(value, weight);
+    }
+
+    @Override
+    public double calculateEntropy()
+    {
+        Map<Double, Double> bucketWeights = Streams.stream(histogram.iterator()).collect(
+                Collectors.groupingBy(
+                        FixedDoubleBreakdownHistogram.BucketWeight::getLeft,
+                        Collectors.summingDouble(e -> e.getCount() * e.getWeight())));
+        double sumW = bucketWeights.values().stream().mapToDouble(Double::doubleValue).sum();
+        if (sumW == 0.0) {
+            return Double.NaN;
+        }
+        long n = Streams.stream(histogram.iterator())
+                .mapToLong(FixedDoubleBreakdownHistogram.BucketWeight::getCount)
+                .sum();
+        double sumWLogW =
+                bucketWeights.values().stream().mapToDouble(w -> w == 0.0 ? 0.0 : w * Math.log(w)).sum();
+
+        double entropy = n * calculateEntropy(histogram.getWidth(), sumW, sumWLogW);
+        for (FixedDoubleBreakdownHistogram.BucketWeight bucketWeight : histogram) {
+            double weight = bucketWeights.get(bucketWeight.getLeft());
+            if (weight > 0.0) {
+                entropy -= getHoldOutEntropy(
+                        n,
+                        bucketWeight.getRight() - bucketWeight.getLeft(),
+                        sumW,
+                        sumWLogW,
+                        weight,
+                        bucketWeight.getWeight(),
+                        bucketWeight.getCount());
+            }
+        }
+        return entropy;
+    }
+
+    private static double getHoldOutEntropy(
```

Comment:
- parameter names here are confusing: bucketWeight and entryWeight, where entryWeight is bucketWeight.getWeight; let's find more descriptive names
- formatting in the body of this function includes unnecessary line breaks

```
    private static double getHoldOutEntropy(
            long n,
            double width,
            double sumW,
            double sumWLogW,
            double bucketWeight,
            double entryWeight,
            long entryMultiplicity)
    {
        double holdoutBucketWeight = Math.max(bucketWeight - entryWeight, 0);
        double holdoutSumW = sumW - bucketWeight + holdoutBucketWeight;
        double holdoutSumWLogW = sumWLogW - getXLogX(bucketWeight) + getXLogX(holdoutBucketWeight);
        return entryMultiplicity * (n - 1) * calculateEntropy(width, holdoutSumW, holdoutSumWLogW) / n;
    }
```